autopush/tests/test_web_base.py

```diff
@@ -0,0 +1,257 @@
+        ch4 = "Access-Control-Expose-Headers"
+        base = self.base
+        base.ap_settings.cors = False
+        assert base._headers.get(ch1) != "*"
+        assert base._headers.get(ch2) != self.CORS_METHODS
+        assert base._headers.get(ch3) != self.CORS_HEADERS
+        assert base._headers.get(ch4) != self.CORS_RESPONSE_HEADERS
+
+        base.clear_header(ch1)
+        base.clear_header(ch2)
+        base.ap_settings.cors = True
+        self.base.prepare()
+        eq_(base._headers[ch1], "*")
+        eq_(base._headers[ch2], self.CORS_METHODS)
+        eq_(base._headers[ch3], self.CORS_HEADERS)
+        eq_(base._headers[ch4], self.CORS_RESPONSE_HEADERS)
+
+    def test_cors_head(self):
+        ch1 = "Access-Control-Allow-Origin"
+        ch2 = "Access-Control-Allow-Methods"
+        ch3 = "Access-Control-Allow-Headers"
+        ch4 = "Access-Control-Expose-Headers"
+        base = self.base
+        base.ap_settings.cors = True
+        base.prepare()
+        base.head(None)
+        eq_(base._headers[ch1], "*")
+        eq_(base._headers[ch2], self.CORS_METHODS)
+        eq_(base._headers[ch3], self.CORS_HEADERS)
+        eq_(base._headers[ch4], self.CORS_RESPONSE_HEADERS)
+
+    def test_cors_options(self):
+        ch1 = "Access-Control-Allow-Origin"
+        ch2 = "Access-Control-Allow-Methods"
+        ch3 = "Access-Control-Allow-Headers"
+        ch4 = "Access-Control-Expose-Headers"
+        base = self.base
+        base.ap_settings.cors = True
+        base.prepare()
+        base.options(None)
+        eq_(base._headers[ch1], "*")
+        eq_(base._headers[ch2], self.CORS_METHODS)
+        eq_(base._headers[ch3], self.CORS_HEADERS)
+        eq_(base._headers[ch4], self.CORS_RESPONSE_HEADERS)
+
+    def test_write_error(self):
+        """ Write error is triggered by sending the app a request
+        with an invalid method (e.g. "put" instead of "PUT").
+        This is not code that is triggered within normal flow, but
+        by the cyclone wrapper.
+        """
+        class testX(Exception):
+            pass
+
+        try:
+            raise testX()
+        except:
+            exc_info = sys.exc_info()
+
+        self.base.write_error(999, exc_info=exc_info)
+        self.status_mock.assert_called_with(999)
+        eq_(self.base.log.failure.called, True)
+
+    def test_write_error_no_exc(self):
+        """ Write error is triggered by sending the app a request
+        with an invalid method (e.g. "put" instead of "PUT").
+        This is not code that is triggered within normal flow, but
+        by the cyclone wrapper.
+        """
+        self.base.write_error(999)
+        self.status_mock.assert_called_with(999)
+        eq_(self.base.log.failure.called, True)
+
+    def test_init_info(self):
+        d = self.base._init_info()
+        eq_(d["request_id"], dummy_request_id)
+        h = self.request_mock.headers
+        h["user-agent"] = "myself"
+        d = self.base._init_info()
```

Comment:
Since this is a hash lookup validation, and you're only testing one key item (without testing to see if other values in the hash are altered) Couldn't you set the mock headers, call `self.base._init_info()` and check the results instead of repeatedly calling `self.base._init_info()`?